### PR TITLE
Add API manager deployment

### PR DESCRIPTION
**Pull request by darkowlzz**

### CODE REVIEW

**Review by croomes — APPROVED** [review]

File: internal/pkg/image/image.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ const (
 	CSIv0ExternalProvisionerContainerImage    = "storageos/csi-provisioner:v0.4.3"
 	CSIv0ExternalAttacherContainerImage       = "quay.io/k8scsi/csi-attacher:v0.4.2"
 	DefaultNFSContainerImage                  = "storageos/nfs:1.0.0"
+	DefaultAPIManagerImage                    = "storageos/api-manager:v0.1.0"
```
Comment (darkowlzz) [inline internal/pkg/image/image.go:21]: Maybe set it to `develop` tag for now, until we have a versioned release. We can update this along with the node container before release.

**Review by darkowlzz — COMMENTED** [review]
Looks great. Left a few comments.

File: pkg/storageos/deploy.go [hunk 2]
```diff
@@ -36,10 +36,12 @@ const (
...
-	daemonsetName   = "storageos-daemonset"
-	statefulsetName = "storageos-statefulset"
-	configmapName   = "storageos-node-config"
-	csiHelperName   = "storageos-csi-helper"
+	daemonsetName         = "storageos-daemonset"
+	statefulsetName       = "storageos-statefulset"
+	configmapName         = "storageos-node-config"
+	csiHelperName         = "storageos-csi-helper"
+	apiManagerName        = "storageos-api-manager"
+	apiManagerMetricsName = "storageos-api-manager-metrics"
```
[the remark this answers is not in the document]
Comment (croomes) [reply pkg/storageos/deploy.go:44]: Moved to api-manager.go and exported, tests updated.

File: pkg/util/k8s/k8s_test.go [hunk 3]
```diff
@@ -186,12 +188,30 @@ func TestResourceManager(t *testing.T) {
...
+		},
 	}
 
 	for _, tc := range testcases {
 		tc := tc
 		t.Run(tc.name, func(t *testing.T) {
-			client := fake.NewFakeClient()
+			s := scheme.Scheme
+			for _, f := range []func(*runtime.Scheme) error{
+				monitoringv1.AddToScheme,
```
[the remark this answers is not in the document]
Comment (croomes) [reply pkg/util/k8s/k8s_test.go:206]: Added & confirmed it fixed the CSI test.

File: internal/pkg/image/image.go (hunk 1 above)
Comment (croomes) [reply internal/pkg/image/image.go]: Pushed `storageos/api-manager:develop`

File: pkg/storageos/deploy_test.go [hunk 4]
```diff
@@ -486,6 +490,113 @@ func TestCreateCSIHelper(t *testing.T) {
...
+				t.Fatal("failed to get the created service", err)
+			}
+
+			// Check the service component label (required for service monitor).
+			got, ok = createdService.Labels[label]
+			if !ok {
+				t.Errorf("expected svc label %q not set", label)
+			}
+			if ok && got != want {
+				t.Errorf("expected svc label %q set to %q, want %q", label, got, want)
```
[the remark this answers is not in the document]
Comment (croomes) [reply pkg/storageos/deploy_test.go]: Fixed

File: pkg/storageos/deploy_test.go [hunk 5]
```diff
@@ -486,6 +490,113 @@ func TestCreateCSIHelper(t *testing.T) {
...
+			if !foundNotReadyTol {
+				t.Errorf("pod toleration %q not found in api manager", nodeNotReadyTolKey)
+			}
+			if !foundUnreachableTol {
+				t.Errorf("pod toleration %q not found for api manager", nodeUnreachableTolKey)
+			}
+
+			// Check the pod template component label (required for service).
+			label := "app.kubernetes.io/component"
+			want := "storageos-api-manager"
```
[the remark this answers is not in the document]
Comment (croomes) [reply pkg/storageos/deploy_test.go]: Done

File: pkg/util/k8s/resource/servicemonitor.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+// Delete deletes a ServiceMonitor.
+func (s ServiceMonitor) Delete() error {
+	return Delete(s.client, getServiceMonitor(s.Name, s.Namespace, s.labels, s.annotations, s.service))
+}
+
+// getServiceMonitor returns a generic ServiceMonitor object.
+func getServiceMonitor(name, namespace string, labels map[string]string, annotations map[string]string, svc *corev1.Service) *monitoringv1.ServiceMonitor {
+	boolTrue := true
+	ownerRef := metav1.OwnerReference{
```
[the remark this answers is not in the document]
Comment (croomes) [reply pkg/util/k8s/resource/servicemonitor.go:71]: Done

File: test/e2e/util/cluster.go [hunk 7]
```diff
@@ -390,3 +391,106 @@ func CSIDriverResourceTest(t *testing.T, driverName string) {
...
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("timed out waiting for api-manager pods: %v", err)
+	}
+
+	// check label needed for metrics service.
+	for _, pod := range pods.Items {
+		label := "app.kubernetes.io/component"
```
[the remark this answers is not in the document]
Comment (croomes) [reply test/e2e/util/cluster.go]: Done

File: test/e2e/util/cluster.go [hunk 8]
```diff
@@ -390,3 +391,106 @@ func CSIDriverResourceTest(t *testing.T, driverName string) {
 		t.Errorf("CSIDriver not found: %v", err)
 	}
 }
+
+// APIManagerDeploymentTest checks the api-manager deployment.
+func APIManagerDeploymentTest(t *testing.T, ns string, retryInterval, timeout time.Duration) {
+	f := framework.Global
+	var deployment *appsv1.Deployment
+	err := wait.Poll(retryInterval, timeout, func() (done bool, err error) {
```
[the remark this answers is not in the document]
Comment (croomes) [reply test/e2e/util/cluster.go]: Done

**Review by darkowlzz — COMMENTED** [review]
LGTM!
Thanks.

**Pull request merged**